source/FreeRTOS_TCP_State_Handling_IPv4.c

```diff
@@ -166,45 +168,62 @@ FreeRTOS_Socket_t * prvHandleListen_IPV4( FreeRTOS_Socket_t * pxSocket,
+            /* MISRA Ref 11.3.1 [Misaligned access] */
+            /* More details at: https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/blob/main/MISRA.md#rule-113 */
+            /* coverity[misra_c_2012_rule_11_3_violation] */
+            const ProtocolHeaders_t * pxProtocolHeaders = ( ( const ProtocolHeaders_t * )
+                                                            &( pxNetworkBuffer->pucEthernetBuffer[ ipSIZE_OF_ETH_HEADER + uxIPHeaderSizePacket( pxNetworkBuffer ) ] ) );
 
-        /* The endpoint in network buffer must be valid in this condition. */
-        pxReturn->pxEndPoint = pxNetworkBuffer->pxEndPoint;
-        pxReturn->bits.bIsIPv6 = pdFALSE_UNSIGNED;
-        pxReturn->u.xTCP.usRemotePort = FreeRTOS_htons( pxTCPPacket->xTCPHeader.usSourcePort );
-        pxReturn->u.xTCP.xRemoteIP.ulIP_IPv4 = FreeRTOS_htonl( pxTCPPacket->xIPHeader.ulSourceIPAddress );
-        pxReturn->u.xTCP.xTCPWindow.ulOurSequenceNumber = ulInitialSequenceNumber;
+            /* The endpoint in network buffer must be valid in this condition. */
+            pxReturn->pxEndPoint = pxNetworkBuffer->pxEndPoint;
+            pxReturn->bits.bIsIPv6 = pdFALSE_UNSIGNED;
+            pxReturn->u.xTCP.usRemotePort = FreeRTOS_htons( pxTCPPacket->xTCPHeader.usSourcePort );
+            pxReturn->u.xTCP.xRemoteIP.ulIP_IPv4 = FreeRTOS_htonl( pxTCPPacket->xIPHeader.ulSourceIPAddress );
+            pxReturn->u.xTCP.xTCPWindow.ulOurSequenceNumber = ulInitialSequenceNumber;
 
-        /* Here is the SYN action. */
-        pxReturn->u.xTCP.xTCPWindow.rx.ulCurrentSequenceNumber = FreeRTOS_ntohl( pxProtocolHeaders->xTCPHeader.ulSequenceNumber );
-        prvSocketSetMSS( pxReturn );
+            /* Here is the SYN action. */
+            pxReturn->u.xTCP.xTCPWindow.rx.ulCurrentSequenceNumber = FreeRTOS_ntohl( pxProtocolHeaders->xTCPHeader.ulSequenceNumber );
+            prvSocketSetMSS( pxReturn );
 
-        prvTCPCreateWindow( pxReturn );
+            xReturnCreateWindow = prvTCPCreateWindow( pxReturn );
 
-        vTCPStateChange( pxReturn, eSYN_FIRST );
+            /* Did allocating TCP sectors fail? */
+            if( xReturnCreateWindow != pdPASS )
+            {
+                /* Close the socket if it was newly created. */
+                if( xIsNewSocket == pdTRUE )
+                {
+                    vSocketClose( pxReturn );
+                }
 
-        /* Make a copy of the header up to the TCP header.  It is needed later
-         * on, whenever data must be sent to the peer. */
-        if( pxNetworkBuffer->xDataLength > sizeof( pxReturn->u.xTCP.xPacket.u.ucLastPacket ) )
-        {
-            xCopyLength = sizeof( pxReturn->u.xTCP.xPacket.u.ucLastPacket );
-        }
-        else
-        {
-            xCopyLength = pxNetworkBuffer->xDataLength;
-        }
+                pxReturn = NULL;
+                break;
```

Comment:
I will suggest to use if else here instead of do{ ... break ... } while( 0 ). It looks more clear to read for me.